---
title: "SCENIC"
author: "PP"
date: "2025-03-04"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## SCENIC

```{bash}
mkdir data
```

```{r packages echo = FALSE, message = FALSE, warning = FALSE, results = 'hide'}

library(Seurat)
library(tidyverse)
library(Matrix)
library(scales)
library(cowplot)
library(RCurl)
library(stringr)
library(ggpubr)
library(RColorBrewer)
library(clustree)
library(lme4)
library(scuttle)
library(edgeR)
library(scran)
library(gridExtra)
library(plyranges) 
library(rtracklayer)
library(gplots)
library(SeuratDisk)
```

**Ortholog data download**

```{bash}
#Get orthoDB data for birds, chicken and mouse
wget https://data.orthodb.org/v12/download/odb12v0_OG2genes.tab.gz -P data
wget https://data.orthodb.org/v12/download/odb12v0_genes.tab.gz -P data
gunzip data/odb12v0_OG2genes.tab.gz
gunzip data/odb12v0_genes.tab.gz

printf "8839_0\n8996_0\n9054_0\n9031_0\n9606_0\n10090_0" > data/hum_mousspecies.txt

cat data/odb12v0_OG2genes.tab | grep at1549675 | \
  grep -e  8839_0 -e 8996_0 -e 9054_0 -e 9031_0 \
  > data/bird_species_orthos.txt
  
cat data/odb12v0_OG2genes.tab | grep at1549675 | \
  grep -e  8839_0 -e 8996_0 -e 9054_0 -e 9031_0 -e 9606_0 -e 10090_0 \
  > data/hum_mouse_bird_species_orthos.txt
  
awk -F'[:\t]' 'NR==FNR {keep[$1]; next} $1 in keep' data/hum_mouse_bird_species.txt data/odb12v0_genes.tab > data/all_genes.txt


#Download reference genome data from NCBI 
#Mallard
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/015/476/345/GCF_015476345.1_ZJU1.0/GCF_015476345.1_ZJU1.0_genomic.gff.gz -P data
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/015/476/345/GCF_015476345.1_ZJU1.0/GCF_015476345.1_ZJU1.0_genomic.fna.gz -P data

#Pheasant
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/004/143/745/GCF_004143745.1_ASM414374v1/GCF_004143745.1_ASM414374v1_genomic.gff.gz -P data
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/004/143/745/GCF_004143745.1_ASM414374v1/GCF_004143745.1_ASM414374v1_genomic.fna.gz -P data

#Guineafowl 
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/078/875/GCF_002078875.1_NumMel1.0/GCF_002078875.1_NumMel1.0_genomic.gff.gz -P data
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/078/875/GCF_002078875.1_NumMel1.0/GCF_002078875.1_NumMel1.0_genomic.fna.gz -P data

#Chicken 
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/016/699/485/GCF_016699485.2_bGalGal1.mat.broiler.GRCg7b/GCF_016699485.2_bGalGal1.mat.broiler.GRCg7b_genomic.gff.gz -P data
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/016/699/485/GCF_016699485.2_bGalGal1.mat.broiler.GRCg7b/GCF_016699485.2_bGalGal1.mat.broiler.GRCg7b_genomic.fna.gz -P data

```

```{r}
#singularity exec docker://rocker/tidyverse R

if (file.exists("lib/rtracklayer") == F){
  install.packages("BiocManager", lib = './lib')
  BiocManager::install("rtracklayer", lib = './lib')
}

library(tidyverse)
library(rtracklayer, lib = './lib')

ggid <- function(x){
  out <- x[[1]][grep("GeneID", x[[1]])] %>% 
    gsub("GeneID:", "", .)
  if (length(out) == 0){
    return("")
  } else {
    return(out)
  }
}


ortho_file <- read.table("data/all_genes.txt", quote = "",  stringsAsFactors = FALSE, sep = '\t')
colnames(ortho_file) <- c("orthodb_geneid", "orthodb_orgid", "protein_id", "synonyms", "uniprot_id",
                          "ensembl_ids", "NCBI_gid", "description", "coords", "DNA_id", "chromosome")

species_df <- data.frame(orthodb_orgid = c('8839_0', '8996_0', '9054_0', '9031_0', '9606_0', '10090_0'), 
                         species = c("Duck", "Guineafowl", "Pheasant", "Chicken", "Human", "Mouse"), 
                         latin = c("Anas_platyrhynchos", "Numida_meleagris", "Phasianus_colchicus", "Gallus_gallus", "", ""), 
                         ref = c("GCF_015476345.1", "GCF_002078875.1", 
                         "GCF_004143745.1", "GCF_016699485.2", "", ""))

ortho_file <- ortho_file %>% 
  merge(species_df)




bird_OGs <- read.table("data/bird_species_orthos.txt")
colnames(bird_OGs) <- c("Bird_cluster", "orthodb_geneid")
bird_mam_OGs <- read.table("data/hum_mouse_bird_species_orthos.txt")
colnames(bird_mam_OGs) <- c("Vert_cluster", "orthodb_geneid")

ortho_file <- ortho_file %>% 
  merge(bird_OGs, all = T) %>% 
  merge(bird_mam_OGs, all = T)

gffs <- list.files("data", full.names = T, pattern = ".gff")

tmp_ofgff <- list()
for (i in 1:4){
  gff <- gffs[grep(species_df[i,4], gffs)]
  z <- import(gff) %>% as.data.frame()
  z2 <- z %>%  filter(type %in% c("CDS")) %>% 
    dplyr::select(gene, type, Dbxref, protein_id) %>% 
    unique()
  species_of <- ortho_file[ortho_file$species == species_df[i,2],]
  gene_ID_df = data.frame(gff_genes  = z2$gene,
                          protein_id = z2$protein_id, 
                          species = species_df[i,2], 
                          GeneID = lapply(z2$Dbxref, ggid) %>% unlist()) %>% 
    merge(species_of, all.y = T)
  print(dim(species_of))
  print(dim(gene_ID_df))
  
  tmp_ofgff[[i]] <- gene_ID_df
}

OF_GFF <- do.call(rbind, tmp_ofgff) %>% 
  bind_rows(ortho_file[ortho_file$species %in% c("Human", "Mouse"),])


# Chicken_synonoyms <- read.table("data/Gallus_gallus.gene_info",
#                        sep = '\t', header = TRUE, comment.char="", fill = TRUE) %>% 
#   dplyr::select(GeneID, Symbol, Synonyms) %>% 
#   dplyr::rename(sSymbol = Symbol)
# 
# 
# Chicken_synonoyms$GeneID[!(Chicken_synonoyms$GeneID %in% OF_GFF$GeneID)]
# 
# OF_GFFs <- merge(OF_GFF, Chicken_synonoyms, by = 'GeneID')


write.table(OF_GFF, file = "data/OrthoDB_GFF_orthologs.csv")

quit()
```

**TF data download and create cluster-buster motif file from JASPAR data**

```{bash}
pwd
#Download from AnimalTFDB v4.0
#wget https://guolab.wchscu.cn/AnimalTFDB4_static/download/TF_list_final/Gallus_gallus_TF -P data
#wget https://guolab.wchscu.cn/AnimalTFDB4_static/download/TF_list_final/Anas_platyrhynchos_platyrhynchos_TF -P data

# Download from SCENIC+ database
wget https://resources.aertslab.org/cistarget/motif_collections/v10nr_clust_public/v10nr_clust_public.zip -P data
unzip data/v10nr_clust_public.zip
mv v10nr_clust_public data
ls data/v10nr_clust_public/singletons | sed 's/\.cb//g' > data/motifs.lst 

```

**Modify motif annotation for each target species**

```{r}
#singularity exec docker://rocker/tidyverse R

library(tidyverse)

OF_GFF <- read.table(file = "data/OrthoDB_GFF_orthologs.csv")
motif_data <- read.table("data/v10nr_clust_public/snapshots/motifs-v10-nr.chicken-m0.00001-o0.0.tbl", sep = '\t', 
                   header = T, comment.char="")
unique(motif_data$gene_name)  %>% length()

OF_GFF_chicken <- filter(OF_GFF, species == "Chicken" & 
                           (gff_genes %in% motif_data$gene_name)) %>% 
  filter(!is.na(Bird_cluster)) %>% 
    dplyr::select(gff_genes, Bird_cluster, GeneID, species) %>% 
    unique()
unique(OF_GFF_chicken$gff_genes)  %>% length()

species_motif_data <- list()
for (sp in c("Duck", "Guineafowl", "Pheasant")){
  OF_GFF_sp <- filter(OF_GFF, species == sp,
                      Bird_cluster %in% OF_GFF_chicken$Bird_cluster) %>% 
    dplyr::select(gff_genes, Bird_cluster, GeneID, species) %>% 
    unique()
  tmp1 <- merge(OF_GFF_sp, OF_GFF_chicken, by = 'Bird_cluster', suffixes = c("_tmp", "_Chicken"), all = F)
  dup1 <- tmp1$GeneID_Chicken[duplicated(tmp1$GeneID_Chicken)]
  dup2 <- tmp1$GeneID_tmp[duplicated(tmp1$GeneID_tmp)]
  tmp2 <- filter(tmp1, GeneID_Chicken %in% dup1 | 
                  GeneID_tmp %in% dup2) %>% 
    filter(gff_genes_tmp == gff_genes_Chicken)

  mgd <- rbind(tmp2, filter(tmp1, !(GeneID_Chicken %in% dup1) & 
                             !(GeneID_tmp %in% dup2)))
  
  genes <- mgd[,"gff_genes_tmp"]
  names(genes) <- mgd$gff_genes_Chicken
  
  species_motif_data[[sp]] <- motif_data %>% 
    mutate(gene_name = genes[gene_name]) %>% 
    filter(!is.na(gene_name))
  print(length(unique(species_motif_data[[sp]]$gene_name)))
  colnames(species_motif_data[[sp]])[1] <- '#motif_id'
  write.table(species_motif_data[[sp]], file = paste0('data/', sp, "_motifs.tbl"), sep = '\t', col.names = T, quote = F, row.names = F)
  
}

quit()
```

**FASTA_FILE Per gene up/downstream region**

1.  For each gene take either the space upstream of the TSS (with or without introns, exons, UTRS or a downstream region)
2.  Default region size for drosophila is 5kb upstream and full length transcript. For mouse and human it's 10kb up and 10kb downstream of the TSS.
3.  For Chicken, blog post recommended 5kb upstream of all genes

```{bash}
conda create -n agat bioconda::agat
conda activate agat 


gunzip data/*fna.gz
gunzip data/*gff.gz

for ref in GCF_015476345.1 GCF_002078875.1 GCF_004143745.1 
do 
echo $ref

agat_sp_extract_sequences.pl --gff data/${ref}*gff --fasta data/${ref}*fna \
  -t gene --up "5000" --down "1000" --full -o data/${ref}_TSS.fasta
done

gzip data/*fna
gzip data/*gff

conda deactivate

```

**Generating cisTarget motif database**

Before running the next section of pySCENIC we need to generate cisTarget_databases

This is essentially a motif x gene matrix, with a ranking score representing the presence of the motif in the regions associated with the gene.

Following this: <https://github.com/aertslab/create_cisTarget_databases/blob/master/README.md>

This function can be split into sections if memory requirements are restrictive

### 

```{bash, eval = FALSE}

git clone https://github.com/aertslab/create_cisTarget_databases
cd create_cisTarget_databases

# Download precompiled Cluster-Buster binary.
wget https://resources.aertslab.org/cistarget/programs/cbust

# Make downloaded binary executable.
chmod a+x cbust
 
wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/liftOver

# Download bigWigAverageOverBed.
wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bigWigAverageOverBed

# Make downloaded binaries executable.
chmod a+x liftOver bigWigAverageOverBed
   
```

**Create the following singularity recipe** #Save as create_cisTarget_databases/CTD_recipe

```         
BootStrap: docker
From: python:3.10

%post
    # Update package lists and install required dependencies
    apt-get update && apt-get install -y \
        python3-pip \
        python3-dev \
    && rm -rf /var/lib/apt/lists/*

    # Upgrade pip
    pip install --upgrade pip

    # Install required Python packages
    pip install \
        "numpy>=1.22" \
        "pandas>=1.4.1" \
        "pyarrow>=7.0.0" \
        "numba>=0.55.1" \
        "flatbuffers"

%environment
    # Set environment variables
    export PATH="/usr/local/bin:$PATH"
    export PYTHONUNBUFFERED=1
    export PATH=$PWD:$PATH

%runscript
    # Default command when the container is run
    /bin/bash
```

```{bash}

singularity pull docker://frolvlad/alpine-glibc:alpine-3

touch CTD_recipe
singularity build CTD.simg CTD_recipe 

singularity run 
```

```{bash}
ref=ref_x
${create_cistarget_databases_dir}/create_cistarget_motif_databases.py \
    -f data/${ref}_TSS.fasta \
    -M data/v10nr_clust_public/singletons \
    -m data/motifs.lst \
    -o ${ref}_motif_db \
    -t ${nbr_threads}
```

-   -f fasta_file: fasta file with sequence per gene ID

-   -M motif_dir: path to Cluster-Buster motifs

```{bash}
singularity exec CTD.simg python 
```

```{r}
cb_files <- list.files("data/motifs_cb_format/", full.names = F)
motifs_data <- lapply(cb_files, function(x){return(c(x, 
                                                     gsub(">", "", 
                                                          readLines(paste0("data/motifs_cb_format/", x))[1]))
                                                   )
  }) %>% 
  do.call(rbind,.) %>% 
  as.data.frame()

colnames(motifs_data) <- c("cb_motif", "TF")


TFs <- paste(motifs_data$TF, collapse = "|")
match <- OF_GFF[apply(OF_GFF, 1, function(x)(any(x %in% motifs_data$TF))),]

intersect(motifs_data$TF, Chicken_TFs$Symbol)

Chicken_TFs <- read.table('data/Gallus_gallus_TF.txt', sep = '\t', header = TRUE)


```

*Read in R object*

```{r seurat_object_load echo = FALSE, message = FALSE, warning = FALSE,  results = 'hide'}

setwd("~/Documents/Science_Work/PhD/Projects/2023/Variables_rates/Avian_scRNAseq/R_analyses/")

sample_info <- read.csv("data/metadata_full.csv", header = F) 
colnames(sample_info) <- c('sample', "species", "ed", "sex", "folder", 'ref', "mito")
sample_info$stage <- "EARLY"
sample_info$stage[sample_info$ed %in% c("ED24", "ED22", "ED25")] <- "LATE"
markers <- read.csv("data/markers/markers.csv") %>% 
  filter(Use == "Yes")
RDatas <- list.files("data/seur_objs/integrated/", full.names = T, pattern = ".RData")
load("data/seur_obj_celltype_metadata.RData")
i = 11
load(RDatas[[i]])
#seurat_integrated@meta.data <- save_seurats[[i]]
#Idents(seurat_integrated) <- 'celltype'
DefaultAssay(seurat_integrated) <- "RNA"
seurat_integrated <- JoinLayers(seurat_integrated)

```

*Output csv file of gene counts matrix*

```{r csv seurat_out}

csv_file <- seurat_integrated@assays$RNA$counts[,1:200] %>% 
  as.data.frame()

write.table(csv_file, "~/Documents/Science_Work/PhD/Projects/2023/Variables_rates/Avian_scRNAseq/SCENIC/data/seurat.csv", sep = ',', row.names = T, col.names = T, quote = F)

tfs <- read.table(
  "~/Documents/Science_Work/PhD/Projects/2023/Variables_rates/Avian_scRNAseq/SCENIC/data/Gallus_gallus_TF.txt", 
  sep = '\t', header = TRUE) %>%
  filter(Symbol != "") %>% 
  filter(!is.na(Entrez_ID)) %>% 
  mutate(Entrez_ID = paste0("LOC", Entrez_ID)) %>% 
  mutate(Symbol = ifelse(Symbol == "", Entrez_ID, Symbol))

tfs2 <- tfs %>% 
  filter(Symbol %in% rownames(seurat_integrated@assays$RNA))

new_tfs <- intersect(toupper(tfs2$Symbol), rownames(tsv_file))
length(new_tfs)

write.table(new_tfs, "~/Documents/Science_Work/PhD/Projects/2023/Variables_rates/Avian_scRNAseq/SCENIC/data/mm_tfs.txt", sep = '\t', row.names = F, col.names = F, quote = F)

```

```{bash}
cd ~/Documents/Science_Work/PhD/Projects/2023/Variables_rates/Avian_scRNAseq/SCENIC/

docker run -p 8888:8888 --platform linux/amd64 -v ~/Documents/Science_Work/PhD/Projects/2023/Variables_rates/Avian_scRNAseq/SCENIC/data:/data aertslab/pyscenic_scanpy:0.12.1_1.9.1 \
  pyscenic grn \
  -o data/out.csv --num_workers 6 -m grnboost2 \
  -t \
  data/seurat.csv data/mm_tfs.txt
        

```

```{r GRN reading, eval = FALSE}

grn_out <- read.csv("~/Documents/Science_Work/PhD/Projects/2023/Variables_rates/Avian_scRNAseq/SCENIC/data/out.csv", header = T)
head(grn_out)


```
